File: android/src/main/kotlin/com/usercentrics/sdk/flutter/api/UsercentricsBannerProxy.kt
package com.usercentrics.sdk.flutter.api

import com.usercentrics.sdk.BannerSettings
import com.usercentrics.sdk.UsercentricsBanner
import com.usercentrics.sdk.UsercentricsConsentUserResponse

internal interface UsercentricsBannerProxy {
    fun showFirstLayer(
        bannerSettings: BannerSettings?,
        callback: (UsercentricsConsentUserResponse?) -> Unit,
    )

    fun showSecondLayer(
        bannerSettings: BannerSettings?,
        callback: (UsercentricsConsentUserResponse?) -> Unit,
    )
}

internal class UsercentricsBannerProxyImpl(
    private val activityProvider: FlutterActivityProvider,
    // Optional: only Consent-or-Pay first-layer screens ever invoke these, so a plain first-layer
    // call with no such buttons configured never touches this at all.
    private val consentOrPayEventNotifier: ConsentOrPayEventNotifier? = null,
) : UsercentricsBannerProxy {

    override fun showFirstLayer(
        bannerSettings: BannerSettings?,
        callback: (UsercentricsConsentUserResponse?) -> Unit,
    ) {
        val context = activityProvider.provide() ?: return
        UsercentricsBanner(context, bannerSettings).showFirstLayer(
            callback = callback,
            onLoginClicked = { url -> consentOrPayEventNotifier?.onLoginClicked(url) },
            onSubscribeClicked = { url -> consentOrPayEventNotifier?.onSubscribeClicked(url) },
        )
    }

    override fun showSecondLayer(
        bannerSettings: BannerSettings?,
        callback: (UsercentricsConsentUserResponse?) -> Unit,
    ) {
        val context = activityProvider.provide() ?: return
        UsercentricsBanner(context, bannerSettings).showSecondLayer(callback)
    }

}

File: ios/usercentrics_sdk/Sources/usercentrics_sdk/API/UsercentricsBannerProxy.swift
import Foundation
import UsercentricsUI
import UIKit

protocol UsercentricsBannerProxyProtocol {
    func showFirstLayer(bannerSettings: BannerSettings?,
                        completionHandler: @escaping (UsercentricsConsentUserResponse) -> Void)

    func showSecondLayer(bannerSettings: BannerSettings?,
                         completionHandler: @escaping (UsercentricsConsentUserResponse) -> Void)
}

struct UsercentricsBannerProxy: UsercentricsBannerProxyProtocol {

    // Optional: only Consent-or-Pay first-layer screens ever invoke these, so a plain first-layer
    // call with no such buttons configured never touches this at all.
    let consentOrPayEventNotifier: ConsentOrPayEventNotifier?

    init(consentOrPayEventNotifier: ConsentOrPayEventNotifier? = nil) {
        self.consentOrPayEventNotifier = consentOrPayEventNotifier
    }

    func showFirstLayer(bannerSettings: BannerSettings?,
                        completionHandler: @escaping (UsercentricsConsentUserResponse) -> Void) {
        UsercentricsBanner(bannerSettings: bannerSettings).showFirstLayer(onLoginClicked: { [weak consentOrPayEventNotifier] url in
            consentOrPayEventNotifier?.onLoginClicked(url)
        }, onSubscribeClicked: { [weak consentOrPayEventNotifier] url in
            consentOrPayEventNotifier?.onSubscribeClicked(url)
        }) { response in
            completionHandler(response)
        }
    }

    func showSecondLayer(bannerSettings: BannerSettings?,
                         completionHandler: @escaping (UsercentricsConsentUserResponse) -> Void) {
        UsercentricsBanner(bannerSettings:bannerSettings).showSecondLayer() { response in
            completionHandler(response)
        }
    }

}
